ng-dev/misc/new-main-branch/cli.ts

```diff
@@ -0,0 +1,133 @@
+  const config = getConfig([assertValidGithubConfig]);
+  const repoSlug = `${config.github.owner}/${config.github.name}`;
+
+  if (config.github.mainBranchName !== 'main') {
+    error(red('Current project is not part of the default branch renaming.'));
+    return;
+  }
+
+  if (git.hasUncommittedChanges()) {
+    error(red('There are uncommitted changes. Unable to switch to new main branch. Aborting..'));
+    return;
+  }
+
+  if (!hasLocalBranch(git, 'master')) {
+    error(red('Local repository does not have a local branch named `master`. Aborting..'));
+    return;
+  }
+
+  if (hasLocalBranch(git, 'main')) {
+    error(red('Local repository already has a branch named `main`. Aborting..'));
+    return;
+  }
+
+  const remotes = getRemotesForRepo(git);
+  const angularRemotes = Array.from(remotes.entries()).filter((r) => isAngularOwnedRemote(r[1]));
+  const angularRemoteNames = angularRemotes.map(([name]) => name);
+
+  // Usually we expect only a single remote pointing to the Angular-owned Github
+  // repository. If there are more, we will just assume the first one to be the primary.
+  const primaryRemoteName = angularRemoteNames[0];
+
+  if (angularRemoteNames.length === 0) {
+    warn(yellow(`Found no remote in repository that points to the \`${repoSlug}\` repository.`));
+  }
+
+  info('The following steps will be performed:');
+  if (angularRemoteNames.length) {
+    info(`  → Remotes (${angularRemoteNames.join(`, `)}) are refreshed.`);
+  }
+  info(`  → The \`main\` branch is fetched from \`${repoSlug}\`.`);
+  info('  → The new `main` branch is checked out.');
+  if (primaryRemoteName) {
+    info(`  → The new \`main\` branch is linked to the \`${primaryRemoteName}\` remote.`);
+  }
+  info('  → The old `master` branch is deleted or renamed (you will be prompted).');
+  info('  → Remote references to `master` branches in the index are removed.');
+  info('');
+
+  if (!(await promptConfirm('Do you want to continue?'))) {
+    return;
+  }
+
+  // Refreshing Angular-owned remotes.
+  git.run(['fetch', ...angularRemoteNames]);
+
+  // Refresh the remote head to account for the new default branch (i.e. `main`).
+  for (const remoteName of angularRemoteNames) {
+    git.run(['remote', 'set-head', remoteName, '-a']);
+  }
+
+  // Fetch the `main` remote branch and store it locally as `refs/heads/main`.
+  git.run(['fetch', git.getRepoGitUrl(), 'main:main']);
+
+  // Checkout the new main branch so that we can potentially delete it later.
+  git.run(['checkout', 'main']);
+
+  // Ensure the local `main` branch has its remote set to an Angular-remote.
+  if (primaryRemoteName !== undefined) {
+    git.run(['branch', '--set-upstream-to', primaryRemoteName, 'main']);
+  }
+
+  // Delete the old `master` branch if desirable, or preserve it if desired.
+  if (
+    await promptConfirm('Are there changes in your local `master` branch that you want to keep?')
+  ) {
+    git.run(['branch', '-m', 'master', 'old-master']);
```

Comment:
We need to verify that `old-master` is available before we can do this safely.

Its probably overly complicated, but we could prompt for the name they want to use.

Reply:
It's a good point. I added a little safety logic, finding a free branch name